### PR TITLE
idmap: handle llistxattr returning EOPNOTSUPP

### PR DESCRIPTION
**Pull request by hallyn**
When shifting a file into idmapset range, we list the xattrs
so we can restore the filecaps if needed.  However, the llistxattr
system call we use for that can return EOPNOTSUPP.  In that case
we should assume an empty xattr set, rather than return an error
and failing the whole file (and container) uid-shift operation.

Signed-off-by: Serge Hallyn <serge@hallyn.com>

### CODE REVIEW

**Comment by tych0** [comment]
IIRC, this is a kernel bug, right? We're doing llistxattr() on overlayfs+squashfs to generate this, and I didn't see any code paths in the kernel that could trigger that.

**Comment by brauner** [comment]
> IIRC, this is a kernel bug, right? We're doing llistxattr() on overlayfs+squashfs to generate this, and I didn't see any code paths in the kernel that could trigger that.

It can, `overlayfs` forwards the listxattr to `squashfs` and if the `squashfs` filesystem doesn't have xattrs it returns `EOPTNOTSUPP`.

**Comment by brauner** [comment]
```
ovl_listxattr()
-> vfs_listxattr()
     -> squahfs_listxattr()
          {
                   /* check that the file system has xattrs */
                  if (msblk->xattr_id_table == NULL)
                          return -EOPNOTSUPP;
          }
```

**Comment by tych0** [comment]
Right sorry. For our specific squashfs images, they do all have xattrs.

**Comment by hallyn** [comment]
yeah still looking into our side, but in the meantime the lxd behavior seems wrong.

**Comment by brauner** [comment]
> yeah still looking into our side, but in the meantime the lxd behavior seems wrong.

What kernel is this on?

File: shared/idmap/idmapset_linux.go [hunk 1]
```diff
@@ -641,22 +641,27 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
...
+				default:
+					return errors.Wrapf(err, "failed getting caps on %s: %v", path, err)
 				}
 			}
 
 			// Shift owner
 			err = ShiftOwner(dir, path, int(newuid), int(newgid))
 			if err != nil {
-				return err
+				return errors.Wrapf(err, "failed shifting %s %s %d %d: %v", dir, path, int(newuid), int(newgid), err)
```
Comment (tomponline) [inline shared/idmap/idmapset_linux.go:657]: Please can you change this to "Failed..." to be consistent with other LXD error messages. Thanks

File: shared/idmap/idmapset_linux.go [hunk 2]
```diff
@@ -641,22 +641,27 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
...
 				caps, err = GetCaps(path)
-				if err != nil {
-					return err
+				switch err {
+				case syscall.EOPNOTSUPP:
+					logger.Warnf("llistxattr is not supported on %s", path)
+				case nil:
+					// noop
+				default:
+					return errors.Wrapf(err, "failed getting caps on %s: %v", path, err)
```
Comment (tomponline) [inline shared/idmap/idmapset_linux.go:650]: Please can you change this to "Failed..." to be consistent with other LXD error messages. Thanks

**Review by tomponline — CHANGES_REQUESTED** [review]
Just a couple of casing nits.

**Comment by tomponline** [comment]
Also doesn't appear to build correctly:

``
shared/idmap/idmapset_linux.go:612:11: undefined: errors
```

**Comment by hallyn** [comment]
> > yeah still looking into our side, but in the meantime the lxd behavior seems wrong.
> 
> What kernel is this on?

What in the hell..  I replied from github app last night, but it doesn't show up here.

This happens on at least 5.10.135 and the ubuntu 5.15.0-43 kernel.  It's not overlay - the squashfs mount itself returns -95 for a bunch of files, but not all.  It's always the same files on both kernels and across reboots.  I need to look at the squashfs binary format.

**Comment by hallyn** [comment]
> Also doesn't appear to build correctly:
> 
> ```
> shared/idmap/idmapset_linux.go:612:11: undefined: errors
> ```

Yes, sorry - I tested this patch on top of 2d57201c38b4ba1287f643af45e428b09074581b but couldn't directly test it on top of master because golang is crap and 1.16 didn't seem to like lxd's go.mod, balking on its grpc dep.  I'll need to do some more testing (or someone else is free to take over).  I'll convert this to draft in the meantime, and turn it back when it's ready.

**Comment by tomponline** [comment]
> 

Yeah LXD requires >= go 1.18 to build now.

File: shared/idmap/idmapset_linux.go (hunk 1 above)
Comment (hallyn) [reply shared/idmap/idmapset_linux.go]: I updated the error messages, but I'm actually not sure what style you want.  I made them more succinct, not bothering to say 'failed'.  If you're after something different, please let me know.

Comment (tomponline) [reply shared/idmap/idmapset_linux.go]: We generally do "Failed ..." (upper case F). Thanks

**Comment by tych0** [comment]
> This happens on at least 5.10.135 and the ubuntu 5.15.0-43 kernel. It's not overlay - the squashfs mount itself returns -95 for a bunch of files, but not all. It's always the same files on both kernels and across reboots. I need to look at the squashfs binary format.

I think it's a race -- I had already added a check for EOPNOTSUPP in `GetAllXattr`, but the fact that we're still getting it suggests it returns something real the first `llistxattr` in that function, but not the second.

**Comment by hallyn** [comment]
> I think it's a race -- I had already added a check for EOPNOTSUPP in GetAllXattr, but the fact that we're still getting it suggests it returns something real the first llistxattr in that function, but not the second.

I think I might be seeing something different than you were.  I can directly mount the same squashfs on a new host and just run llistxattr.c on one of the -95 paths and still get -95.  This is different from the behavior you're describing.

Now maybe if I can understand why I'm getting the -95, that might also explain the race you were getting... I can only hope.

**Comment by tych0** [comment]
https://github.com/lxc/lxd/pull/10198

**Comment by hallyn** [comment]
> #10198

A-ha - yes, the lxd version we're using here doesn't have that fix yet.  I'll test with just that one, and if that suffices - as it should - i'll close this.  Thanks.

**Comment by hallyn** [comment]
Even so - I stil don't get why, on the same squashfs file, some files return -95 and some return just nil err and sz 0.

**Comment by tych0** [comment]
Yeah... I think there's a bug somewhere. It's probably "fine" to paper over it by ignoring EOPNOTSUPP harder, but it might be interesting to know what causes it.

**Comment by hallyn** [comment]
> > #10198
> 
> A-ha - yes, the lxd version we're using here doesn't have that fix yet. I'll test with just that one, and if that suffices - as it should - i'll close this. Thanks.

Yeah that suffices, so I'll close this, thanks.  (Mind you, the extra debug printing may be useful in pathological cases, but then that should be a separate patch).

Closing - thanks.

**Pull request closed without merging**